### PR TITLE
Split actions into reusable steps (fixes #213, fixes #77, embeds #233, #236, #253)

### LINKED ISSUE #77: How to extend existing actions?
In #75 we are adding more logic to the default action behavior. Namely setting assignee and status fields.

The chosen approach was to extend the default action behavior using inheritance.

**Pros**

- No duplicated code between actions

**Cons**

- Indirection. One has to read and mentally merge two classes in order to understand the final behavior
- Hard to pick (eg. I want status to be set, but not the assignee)

There could be other approaches to consider:

## Single action with (many) parameters

```yaml
    action: src.jbi.whiteboard_actions.default_action
    contact: [example@allizom.com]
    description: example configuration
    enabled: true
    parameters:
      jira_project_key: EXMPL
      whiteboard_tag: example
      sync_assignee: True
      status_map:
        NEW: "In Progress"
        FIXED: "Closed"
```

**Pros**

- A single piece of code to read  (`default_action.py`)

**Cons**

- The default action code could potentially grow indefinitely, and we would loose the extensibility offered by the action API
- Unit tests have to combine a lot of combinations for the parameters

## Support execution of multiple atomic actions

Instead of a single action, define a list of actions to be executed. The specified parameters would be passed to all to simplify configuration.

```yaml
    contact: [example@allizom.com]
    description: example configuration
    enabled: true
    actions:
       - src.jbi.whiteboard_actions.create_or_update
       - src.jbi.whiteboard_actions.comments
       - src.jbi.whiteboard_actions.assignee
       - src.jbi.whiteboard_actions.status
    parameters:
      jira_project_key: EXMPL
      whiteboard_tag: example
      status_map:
        NEW: "In Progress"
        FIXED: "Closed"
```

**Pros**

- Smaller bits of code
- Modularity
- Easier testing

**Cons**

- Potential duplication in config if all use-cases want the same set of actions
- Multiple calls to API for separate fields (verbose history in jira?)

----

There could be other approaches. Maybe the current one is fine, but at least this issue could help us drive a conversation about it (before the code would grow too much)

### LINKED ISSUE #213: Make the `default` action (and the logic contained inside it) easier to reuse
These are some ideas I had to make reusing logic present in the default action easier:

#### Break up long methods so that they have fewer responsibilities

The goal here would be to make methods easier to override while still allowing action authors to reuse sensible default logic.

For example, the `create_and_link_issue` method:
- creates an issue on Jira
- checks if that issue is a duplicate. If is isn't:
- links the Jira ticket to bugzilla
- appends the Jira issue URL to the bugzilla bug

If an author wanted to do something else during the `create_and_link_issue` method, they'd probably have to copy + paste the whole method just to add their logic.

But if `create_and_link_issue` looked more like:

```python
def create_and_link_issue(self, *args, **kwargs):
    issue = self.create_issue(...)
    is_dupe = self.check_if_duplicate(...)
    if is_dupe:
        return self.rollback_issue_creation(...)
    link_bug_response = self.link_bug_to_issue(...)
    link_issue_response = self.link_issue_to_bug(...)
    return True, {...}
```

it would be easier for an author to adapt this method to suit their needs, either by:
```python
def create_and_link_issue(self, *args, **kwargs):
    response = super().create_and_link_issue(self, *args, **kwargs)
    additional_response = self.my_additional_thing(...)
    return True, {...}
```
or 

```python
def create_and_link_issue(self, *args, **kwargs):
    issue = self.create_issue(...)
    # we don't care about dupes
    link_bug_response = self.link_bug_to_issue(...)
    link_issue_response = self.link_issue_to_bug(...)
    additional_response = self.my_additional_thing(...)
    return True, {...}
```

#### Move more functions / methods to service module

Some of the logic we currently have in the default action would probably be better suited for the service module. This would help slim down methods in the default action, which would make reusing it easier. It would also make creating brand new actions easier.

These methods would probably know how to receive our models (mostly payloads) and do things with them. For instance:
```python
def create_jira_issue_from_payload(payload, *args, **kwargs):
def update_assignee(payload, *args, **kwargs):
def update_status(payload, *args, **kwargs):
```

### PR DESCRIPTION
**Pull request by leplatrem**
fixes #213

I removed some of the class-based stuff, but not everything, we should agree on a design for #77 before we can go further IMO

### CODE REVIEW

**Review by bsieber-mozilla — APPROVED** [review]
Thanks for the efforts in this PR and #233.

I think we can land this and iterate on this "pipeline" approach we discussed in the meeting today.

**Review by grahamalama — APPROVED** [review]
Looks good! I'm approving this knowing that #233 is a fast follow and we'll be quickly iterating on this design, but these were just things I wanted to comment on for posterity if for nothing else.

File: jbi/actions/default.py [hunk 1]
```diff
@@ -45,249 +52,318 @@ def __init__(self, jira_project_key, **kwargs):
         self.jira_project_key = jira_project_key
         self.sync_whiteboard_labels = kwargs.get("sync_whiteboard_labels", True)
 
-        self.bugzilla_client = bugzilla.get_client()
-        self.jira_client = jira.get_client()
```
[the remark this answers is not in the document]
Comment (leplatrem) [reply jbi/actions/default.py:49]: Yes, I agree, composition pattern is better 👍

File: jbi/actions/default.py [hunk 2]
```diff
@@ -45,249 +52,318 @@ def __init__(self, jira_project_key, **kwargs):
...
-        return fields
+        self.on_create_issue(log_context, bug=bug, event=event, issue_key=issue_key)
+
+        return True, {
+            "bugzilla_response": bugzilla_response,
+            "jira_response": jira_response,
+        }
 
-    def jira_comments_for_update(
+    def on_create_issue(
```
[the remark this answers is not in the document]
Comment (leplatrem) [reply jbi/actions/default.py:166]: Your feedback on this is totally relevant.

I wasn't a big fan either, but used it as a kind of segway from the previous approach.

**Comment by grahamalama** [comment]
> Maybe we should cut a release before landing this beast on main though

That's a good idea!

**Comment by grahamalama** [comment]
Before we merge this, could we update the title / description to reflect the fact that it's an "omnibus" PR that includes #233 and #253?

**Pull request merged**